chain/events/filter/index_test.go

```diff
@@ -957,3 +962,84 @@ func TestEventIndexPrefillFilterExcludeReverted(t *testing.T) {
 		})
 	}
 }
+
+// TestQueryPlan is to ensure that future modifications to the db schema, or future upgrades to
+// sqlite, do not change the query plan of the prepared statements used by the event index such that
+// queries hit undesireable indexes which are likely to slow down the query.
+// Changes that break this test need to be sure that the query plan is still efficient for the
+// expected query patterns.
+func TestQueryPlan(t *testing.T) {
+	dbPath := filepath.Join(t.TempDir(), "actorevents.db")
+
+	ei, err := NewEventIndex(context.Background(), dbPath, nil)
+	require.NoError(t, err, "create event index")
+
+	rv := reflect.ValueOf(ei).Elem()
+	rt := rv.Type()
+
+	for i := 0; i < rv.NumField(); i++ {
+		field := rv.Field(i)
+		fieldType := rt.Field(i)
+
+		// Check if the field is unexported and a string
+		if fieldType.PkgPath != "" && strings.HasPrefix(fieldType.Name, "stmt") {
+			fmt.Printf("Found unexported string field: %v, Value: %v\n", fieldType.Name, field.String())
+		}
+	}
+
+	verifyQueryPlan := func(stmt string) {
+		rows, err := ei.db.Query("EXPLAIN QUERY PLAN " + strings.Replace(stmt, "?", "1", -1))
+		require.NoError(t, err, "explain query plan for query: "+stmt)
+		defer func() {
+			require.NoError(t, rows.Close())
+		}()
+		// First response to EXPLAIN QUERY PLAN should show us the use of an index that we want to
+		// encounter first to narrow down the search space - either a height or tipset_key_cid index
+		// - sqlite_autoindex_events_seen_1 is for the UNIQUE constraint on events_seen
+		// - events_seen_height and events_seen_tipset_key_cid are explicit indexes on events_seen
+		// - event_height and event_tipset_key_cid are explicit indexes on event
+		rows.Next()
+		var id, parent, notused, detail string
+		require.NoError(t, rows.Scan(&id, &parent, &notused, &detail), "scan explain query plan for query: "+stmt)
+		require.Regexp(t, regexp.MustCompile(" INDEX (sqlite_autoindex_events_seen_1|event(s_seen)?_(height|tipset_key_cid))( |$)"), detail, "index used for query: "+stmt)
+		t.Logf("[%s] has plan start: %s", strings.Replace(regexp.MustCompile(`(?m)^\s+`).ReplaceAllString(stmt, " "), "\n", "", -1), detail)
```

Comment:
Idea (can skip): In case this helps others, add comment like "removes leading whitespace and newlines from stmt"

Reply:
yeah, it was more explicit originally but I decided to compact it because it's just a log line; I've reverted it now, I agree on being more explicit